### PR TITLE
Require GetApplicationBoxes to return all Box names

### PR DESCRIPTION
**Pull request by michaeldiamant**
Attempts to address https://github.com/algorand/go-algorand/pull/4149#discussion_r1001083810 by modifying _GetApplicationBoxes_ to fail when either a client or server-side response size limit _prevents_ returning all Box names.  With the PR, requests succeed _only_ when all Box names can be returned.

### CODE REVIEW

File: daemon/algod/api/server/v2/handlers.go [hunk 1]
```diff
@@ -1219,24 +1219,24 @@ func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64,
...
 
-	// cast param.Max to zero, if param.Max points to some value, we keep the value
-	castedMax := nilToZero(params.Max)
-	maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication
-	algodLimitsResults := maxBoxThreshold > 0
-
-	boxKeys, err := ledger.LookupKeysByPrefix(lastRound, keyPrefix, applicationBoxesMaxKeys(castedMax, maxBoxThreshold))
+	requestedMax, algodMax := nilToZero(params.Max), v2.Node.Config().MaxAPIBoxPerApplication
+	max := applicationBoxesMaxKeys(requestedMax, algodMax)
+	record, _, _, err := ledger.LookupAccount(ledger.Latest(), appIdx.Address())
```
[the remark this answers is not in the document]
Comment (michaeldiamant) [reply daemon/algod/api/server/v2/handlers.go]: @cce https://github.com/algorand/go-algorand/pull/4689/commits/bf11eebca1988b52ff58c606914ee8da72197b99 makes the requested change.  If it looks good to you, can you resolve the thread?

File: daemon/algod/api/server/v2/handlers.go [hunk 2]
```diff
@@ -1219,24 +1219,24 @@ func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64,
...
 		return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
 	}
-
-	if algodLimitsResults && uint64(len(boxKeys)) > maxBoxThreshold {
-		v2.Log.Info("MaxAPIBoxPerApplication limit %d exceeded", maxBoxThreshold)
+	if record.TotalBoxes > max {
 		return ctx.JSON(http.StatusBadRequest, generated.ErrorResponse{
-			Message: "Result limit exceeded",
+			Message: fmt.Sprintf("Result limit exceeded: MaxAPIBoxPerApplication=%d, max=%d, TotalBoxes=%d", algodMax, requestedMax, record.TotalBoxes),
 			Data:    nil,
```
[the remark this answers is not in the document]
Comment (michaeldiamant) [reply daemon/algod/api/server/v2/handlers.go]: @cce It's a good callout.  https://github.com/algorand/go-algorand/pull/4689/commits/bf11eebca1988b52ff58c606914ee8da72197b99 and https://github.com/algorand/go-algorand/pull/4689/commits/fd6215b201505d031f176cdc6fe2339699bd48d9 amend the response structure to be JSON.

**Review by jannotti — APPROVED** [review]
I'm ok with it as is, but I think I'd rather.
a) not take a max parameter at all
b) remove max handling from the internal ledger interface

Both would simplify code and more closely resemble the way it was done for assets/apps.

**Comment by michaeldiamant** [comment]
@jannotti Merging so that we can get closer to the desired behavior.  I'm open to making more changes - I left thoughts below.  Hearing nothing, I'll assume we're _OK_ with the implemented behavior.

> a) not take a max parameter at all

I considered it, but chose to keep as is for several reasons:
* It feels to me like the caller ought to have _some_ control over response size.  
* Not the strongest rationale - There's downstream changes needed for any change here.

> b) remove max handling from the internal ledger interface

I initially made changes removing the max handling.  I left it in place assuming it's otherwise desirable and because there's additional instances where queries are limited by a max.  I don't feel too strongly here.

**Pull request merged**